src/Client/BuzzHouse/Generator/SQLTable.cpp

```diff
@@ -705,6 +720,18 @@ const DB::Strings & s3_compress = {"none", "gzip", "gz", "brotli", "br", "xz", "
 void StatementGenerator::generateEngineDetails(RandomGenerator & rg, SQLBase & b, const bool add_pkey, TableEngine * te)
 {
     SettingValues * svs = nullptr;
+    const bool has_tables = collectionHas<SQLTable>(
+        [&](const SQLTable & t)
+        {
+            return t.db && t.db->attached == DetachStatus::ATTACHED && t.attached == DetachStatus::ATTACHED
+                && (t.is_deterministic || !b.is_deterministic);
+        });
+    const bool has_views = collectionHas<SQLView>(
+        [&](const SQLView & v)
+        {
+            return (!v.db || v.db->attached == DetachStatus::ATTACHED) && v.attached == DetachStatus::ATTACHED
+                && (v.is_deterministic || !b.is_deterministic);
+        });
```

Comment:
I don't understand why you have this same logic so many times. In this file and in StatementGenerator. What's against moving them to functions that can be reused anywhere? Even if there are some extra checks that need to be done in some cases, you can still reuse the same common part and add extra statements on top in a new lambda.

Reply:
Fixed